### ISSUE TITLE
RTK Query - refetchOnFocus refetches unmounted queries

### ISSUE THREAD

**Issue by saschametz** [reduxjs/redux-toolkit#1530]
We're using the [`refetchOnFocus`](https://redux-toolkit.js.org/rtk-query/api/createApi#refetchonfocus) option and set [`keepUnusedDataFor`](https://redux-toolkit.js.org/rtk-query/api/createApi#keepunuseddatafor) to 5 minutes.

The problem is that when the user navigates through the application, this creates a lot of API requests at once when the page is refocused.

This question already seems to have been up for discussion in the old repo:
- https://github.com/rtk-incubator/rtk-query/pull/90#discussion_r540368910
- https://github.com/rtk-incubator/rtk-query/issues/99
---
It would be very useful to change the default behaviour of this (refetch mounted queries only on refocus) or have an option to disable the fetching of unmounted queries.

Changing the default behaviour would be the preferable way IMHO. We don't have the problem of stale data for unmounted queries because we're using the [`refetchOnMountOrArgChange`](https://redux-toolkit.js.org/rtk-query/api/createApi#refetchonmountorargchange) option.

**Comment by phryneas** [comment]
Hmm, the questions is really what the best - most correct - behaviour would be. If it's not refetching that data, it might grow stale and not everybody uses `refetchOnMountOrArgChange`.

The "cleanest" way I see for this would be to just remove that stale data from the store, but that would maybe not be desirable for everyone as well :thinking:.

Maybe we could make it configurable between
* keep
* refetch
* delete
?

**Comment by saschametz** [comment]
I'm used to react-query where it only fetches the mounted queries on refocus. But I see this may not be the best solution because `refetchOnMountOrArgChange` is not active by default.

I was actually surprised at the number of queries on refocus. I just spotted it by accident and I think this can happen to others as well, that they don't even know that they may produce a lot of queries by using `refetchOnFocus`. Would it be an option to change the default value of `refetchOnMountOrArgChange` to `true` so that the default behavior of `refetchOnFocus` refetches only mounted queries?

Having the option between keeping the stale data or refetching it for `refetchOnFocus` sounds good to me. I'm not sure about the delete option, though, because if you click on another tab for just one second, the cached data is already deleted.

**Comment by AlexanderArvidsson** [comment]
I am also noticing this problem and after reading https://github.com/reduxjs/redux-toolkit/issues/1409 it seems like this logic observed here is contradictory to the arguments made in that issue.

In that issue, it is mentioned that revalidation only occurs if there are active subscribers to the data. However, here, when focusing, it will refetch ALL endpoints regardless if there are any subscribers.

I am using the following options:
```
  refetchOnMountOrArgChange: 1,
  refetchOnFocus: true,
  refetchOnReconnect: true,
```
I am seeing LOTS of executeQuery actions for endpoints that clearly have no subscribers. Looking in the state before focusing the window, the majority of endpoints have no subscribers. When refocusing, all endpoints are refetched.

I am noticing, however, that this might actually be a bug.
After leaving the page that initiates certain queries and removing the active subscribers, a part of my state looks like this:
![image](https://user-images.githubusercontent.com/2972103/148377469-eef77dad-4fbb-4094-9c63-35a5a2445112.png)

If I refresh the page,

**Comment by AlexanderArvidsson** [comment]
Any progress on this?

This seems like it's a critical bug as it is refetching queries that have no active subscriptions **which is really should not**.

**Comment by phryneas** [comment]
Sorry, when you left your last comment I had Covid and I haven't cought up on everything yet.

Generally, just "not refetching" doesn't really cut it here, as in that case the data has to be completely removed from the cache, so it is re-queried on a remount. Similar logic to invalidation would apply here.
Unfortunately I won't get to it in quite some while as I still have real life to catch up, but it would be great if you could open a PR on this.

**Comment by AlexanderArvidsson** [comment]
@phryneas I opened a PR for the simple change of adding the condition, but if that is not enough I would need some more information to fix it completely.

I have not dove too far into the code for RTK Query, but enough to see roughly how it works. 
So if I understand you correctly, in order for the data to be re-queried when mounting the component, the data has to be removed from the cache, else it will have stale data?

However, wouldn't the subscription state become non-empty once you mount your component again, resulting in the if statement not passing and thus refetching the data correctly?

**Comment by phryneas** [comment]
It would only refetch if `refetchOnMountOrArgChange` is set - if that is not set it will reuse the existing cache value on mount. So if all other data was refreshed, but this one cache value was not, a newly mounting component will get a cache value that is potentially outdated.

Essentially the same as with invalidation goes: if it is mounted, it needs a refetch, if it is not mounted, it needs to be cleaned out so it is fetched on a later mount in the future.
https://github.com/reduxjs/redux-toolkit/blob/d4e95caa26ea1c76a1ecead8f09b1897b4444d40/packages/toolkit/src/query/core/buildMiddleware/invalidationByTags.ts#L75-L82

**Comment by AlexanderArvidsson** [comment]
@phryneas Isn't this already handled by the `handleUnsubscribe` though?

https://github.com/reduxjs/redux-toolkit/blob/d5b0eb59ee6279bbc1e63000655f3432056f06d2/packages/toolkit/src/query/core/buildMiddleware/cacheCollection.ts#L79-L103

If you have `keepUnusedDataFor` enabled, it should clear the query data after the time has passed, at which point a remount will query the data again (regardless if `refetchOnMountOrArgChange` is set, because the query data is already cleaned up)

If you do not have `keepUnusedDataFor`, `handleSubscribe` will remove the data right as you unsubscribe (i.e. unmount) at which point a remount will also query the data again.

If you have `refetchOnMountOrArgChange` enabled, it will always requery when remounting.

In all these cases, the fix in my PR works as intended. Have I missed something?

**Comment by phryneas** [comment]
Case 4 is the default case. `keepUnusedDataFor` defaults to 60 seconds and `refetchOnMountOrArgChange ` defaults to `false`. It will probably be the 90% use case.

"serving a cached result" is different from "serving something we know to be out of sync with all other cache results". If you use `refetchOnFocus` you expect all data to be up-to-date after refocus. You would not expect data to stay around for longer and be older than the last refocus when a component mounts with a value that was available in cache.

Thus, that data has to be removed from cache in that case.

**Comment by AlexanderArvidsson** [comment]
In my use case, I am using `refetchOnFocus` (throttled, though, custom handler), `refetchOnMountOrArgChange` and `keepUnusedDataFor = 60` (default). 

If I initiate a query and unmount the component, then blur the window and refocus it, it will re-query all data again for the component that was unmounted (no component is listening on the data).

If the reasoning is that `refetchOnFocus` is meant to refetch _all_ data, including _unused_ data, then I understand the current logic. 

However, my question is then, should we ignore `keepUnusedDataFor` and remove the query data if there are no subscribers when trying to refetch?

**Comment by phryneas** [comment]
> However, my question is then, should we ignore keepUnusedDataFor and remove the query data if there are no subscribers when trying to refetch?

Yes, it is the same we do for `invalidate` with `keepUnusedDataFor`. At that point we have to assume the data in the cache is outdated - or at least not "in sync" with other stuff we are displaying.
So we can either refetch it, which can be costly for stuff we maybe never display again. Or we remove it and it will be refetched if it is actually required. Either way, we cannot keep around a potentially outdated version of that data.

So I guess that the current behaviour might be overfetching a lot, so at this point I'm leaning towards removing it. We could also think about making that configurable and a user choice, but that might be going too far.

**Comment by AlexanderArvidsson** [comment]
> PS: "data older than keepUnusedDataFor" will never happen. That data will be removed once it hits that point.

Yes, of course, I was not referring to it as actually existing, it is removed exactly on time due to the setTimeout.

I am beginning to understand exactly what you are saying and I agree, I have implemented the same logic from `invalidateTags` and it works fine for my use case.

To be exact with my use case, I am calling the query without hooks (as per documentation) and instantly unsubscribing (I only want the data, I do not want to subscribe). This is because I am storing the results in an entity adaptor that I want to manage manually (experimenting with infinite pagination). The problem then arose where it started refetching my data even though I did subscribe and that was unnecessary (we have real-time long polling that I will integrate into the entity adaptor to keep in sync). Removing the data like `invalidateTags` works very well in this case.

I will update my PR with these changes, but first I will also go to bed.
Good talk and appreciate the fast responses!
